File: compilers/python/runar_compiler/frontend/diagnostic.py
"""Structured compiler diagnostics."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from runar_compiler.frontend.ast_nodes import SourceLocation


class Severity:
    ERROR = "error"
    WARNING = "warning"


@dataclass
class Diagnostic:
    """A single compiler diagnostic (error or warning)."""
    message: str
    severity: str  # Severity.ERROR or Severity.WARNING
    loc: Optional[SourceLocation] = None

    def format_message(self) -> str:
        """Format with optional file:line:column prefix."""
        if self.loc and self.loc.file:
            if self.loc.column > 0:
                return f"{self.loc.file}:{self.loc.line}:{self.loc.column}: {self.message}"
            return f"{self.loc.file}:{self.loc.line}: {self.message}"
        return self.message

    def __str__(self) -> str:
        return self.format_message()

File: compilers/rust/src/frontend/diagnostic.rs
//! Structured compiler diagnostics.
//!
//! Provides a `Diagnostic` type that mirrors the TypeScript compiler's
//! `CompilerDiagnostic` type: `{ message, loc?, severity }`.

use super::ast::SourceLocation;
use std::fmt;

/// Severity levels for compiler diagnostics.
#[derive(Debug, Clone, PartialEq)]
pub enum Severity {
    Error,
    Warning,
}

/// A single compiler diagnostic (error or warning).
#[derive(Debug, Clone)]
pub struct Diagnostic {
    pub message: String,
    pub loc: Option<SourceLocation>,
    pub severity: Severity,
}

impl Diagnostic {
    /// Create a new diagnostic.
    pub fn new(message: impl Into<String>, severity: Severity, loc: Option<SourceLocation>) -> Self {
        Self {
            message: message.into(),
            loc,
            severity,
        }
    }

    /// Create an error diagnostic.
    pub fn error(message: impl Into<String>, loc: Option<SourceLocation>) -> Self {
        Self::new(message, Severity::Error, loc)
    }

    /// Create a warning diagnostic.
    pub fn warning(message: impl Into<String>, loc: Option<SourceLocation>) -> Self {
        Self::new(message, Severity::Warning, loc)
    }

    /// Format the diagnostic with optional file:line:column prefix.
    pub fn format_message(&self) -> String {
        if let Some(ref loc) = self.loc {
            if !loc.file.is_empty() {
                if loc.column > 0 {
                    return format!("{}:{}:{}: {}", loc.file, loc.line, loc.column, self.message);
                }
                return format!("{}:{}: {}", loc.file, loc.line, self.message);
            }
        }
        self.message.clone()
    }
}

impl fmt::Display for Diagnostic {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.format_message())
    }
}
